flink-tests/src/test/java/org/apache/flink/test/scheduling/PipelinedRegionSchedulingITCase.java

```diff
@@ -0,0 +1,130 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.test.scheduling;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.client.program.MiniClusterClient;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.JobManagerOptions;
+import org.apache.flink.configuration.RestOptions;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionType;
+import org.apache.flink.runtime.jobgraph.DistributionPattern;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobgraph.JobVertex;
+import org.apache.flink.runtime.jobgraph.ScheduleMode;
+import org.apache.flink.runtime.jobmanager.scheduler.NoResourceAvailableException;
+import org.apache.flink.runtime.jobmanager.scheduler.SlotSharingGroup;
+import org.apache.flink.runtime.jobmaster.JobResult;
+import org.apache.flink.runtime.minicluster.MiniCluster;
+import org.apache.flink.runtime.minicluster.MiniClusterConfiguration;
+import org.apache.flink.runtime.testtasks.NoOpInvokable;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * IT case for pipelined region scheduling.
+ */
+public class PipelinedRegionSchedulingITCase extends TestLogger {
+
+	@Test
+	public void testSuccessWithSlotsNoFewerThanTheMaxRegionRequired() throws Exception {
+		final JobResult jobResult = executeSchedulingTest(2);
+		assertThat(jobResult.getSerializedThrowable().isPresent(), is(false));
+	}
+
+	@Test
+	public void testFailsOnInsufficientSlots() throws Exception {
+		final JobResult jobResult = executeSchedulingTest(1);
+		assertThat(jobResult.getSerializedThrowable().isPresent(), is(true));
+
+		final Throwable jobFailure = jobResult
+			.getSerializedThrowable()
+			.get()
+			.deserializeError(ClassLoader.getSystemClassLoader());
+
+		final Optional<NoResourceAvailableException> cause = ExceptionUtils.findThrowable(
+			jobFailure,
+			NoResourceAvailableException.class);
+		assertThat(cause.isPresent(), is(true));
+		assertThat(cause.get().getMessage(), containsString("Slot request bulk is not fulfillable!"));
+	}
+
+	private JobResult executeSchedulingTest(int numSlots) throws Exception {
+		final Configuration configuration = new Configuration();
+		configuration.setString(RestOptions.BIND_PORT, "0");
+		configuration.setLong(JobManagerOptions.SLOT_REQUEST_TIMEOUT, 5000L);
```

Comment:
The JM needs to wait RM connected and TM registered before it can obtain slot. So a timeout is needed otherwise the case `testSuccessWithSlotsNoFewerThanTheMaxRegionRequired` can fail (e.g. setting the timeout to 10ms the test almost fails every time).
The larger timeout, the less chance it will fail. I think 5s is not pain for such kind of tests.

Reply:
The main concern is the time to build the connections between JM and TMs and slot offers. It can happen only after a job is launched.

Reply:
ok